webapp/src/main/java/org/opentestsystem/rdw/admin/service/impl/DefaultStudentGroupBatchService.java

```diff
@@ -65,43 +66,41 @@ public DefaultStudentGroupBatchService(@NotNull final StudentGroupBatchRepositor
 
     @Override
     public StudentGroupBatch upload(final User user, final MultipartFile file) {
+        logger.debug("Processing upload request: {}", file.getOriginalFilename());
 
-        final StudentGroupBatch.Builder studentGroupBatchBuilder = StudentGroupBatch.builder().creator(user.getUsername());
-        StudentGroupBatch sgbImport;
+        final PermissionScope permissionScope = user
+                .getPermissionsById()
+                .get("GROUP_WRITE")
+                .getScope();
 
-        final MessageDigest messageDigest = DigestUtils.getMd5Digest();
-        try (final DigestInputStream digestStream = new DigestInputStream(file.getInputStream(), messageDigest)) {
-            validate(digestStream);
-        } catch (final Exception ex) {
-            studentGroupBatchBuilder
-                    .digest("") //TODO: should digest be optional?
-                    .message(ex.getMessage())
-                    .status(ImportStatus.BAD_DATA);
-            return repository.create(studentGroupBatchBuilder.build());
-        }
+        final StudentGroupBatch.Builder studentGroupBatchBuilder = StudentGroupBatch.builder()
+                .creator(user.getUsername());
 
-        final String digest = Hex.encodeHexString(messageDigest.digest()).toUpperCase();
-        try (final InputStream fileStream = file.getInputStream()) {
-            final Properties properties = new Properties();
-            properties.put(Headers.CONTENT_TYPE, file.getContentType());
-            properties.put(Headers.CONTENT_LENGTH, file.getSize());
-            properties.setProperty("username", user.getUsername());
-            properties.setProperty("filename", file.getName());
+        final MessageDigest messageDigest = DigestUtils.getMd5Digest();
+        final List<CsvValidationFailure> validationFailures = new ArrayList<>();
+        try {
+            try (final DigestInputStream digestStream = new DigestInputStream(file.getInputStream(), messageDigest)) {
+                validationFailures.addAll(validate(digestStream, permissionScope));
+            }
 
-            //TODO: check if the digest already exists
-            sgbImport = repository.findOneByDigest(digest);
-            if (sgbImport != null) {
-                logger.info("Ignoring file name {} with existing digest {}", file.getName(), digest);
+            final String digest = Hex.encodeHexString(messageDigest.digest()).toUpperCase();
+            if (!validationFailures.isEmpty()) {
+                studentGroupBatchBuilder
+                        .digest(digest)
+                        .message(validationService.toMessage(validationFailures))
+                        .status(ImportStatus.BAD_DATA);
+                logger.debug("Request failed validation: {}", file.getOriginalFilename());
+                return repository.create(studentGroupBatchBuilder.build());
```

Comment:
Then we'd have to shove everything below into an else block (or additional if) since we don't want to modify the StudentGroupBatchBuilder or attempt to archive the file if the upload failed validation.